File: watchOS3/HeartRateWatchSample/HeartRateWatchSample WatchKit Extension/InterfaceController.m
//
//  InterfaceController.m
//  HeartRateWatchSample WatchKit Extension
//
//  Created by Luis Castillo on 10/4/16.
//  Copyright © 2016 LC. All rights reserved.
//

#import "InterfaceController.h"


@interface InterfaceController()
{
    WorkoutSessionService * wkService;
    NSDateFormatter * dateF;
}
@end


@implementation InterfaceController

@synthesize wkStartDateLabel, wkEndDateLabel, wkHrValueLabel;
@synthesize startWkButton, endWkButton;

- (void)awakeWithContext:(id)context {
    [super awakeWithContext:context];

    //UI setup
    [wkStartDateLabel setText:@"Start: "];
    [wkEndDateLabel setText:@"End:  "];
    [wkHrValueLabel setText:@" . . . "];

    //formatter
    dateF = [[NSDateFormatter alloc]init];
    [dateF setDateStyle:NSDateFormatterNoStyle];
    [dateF setTimeStyle:NSDateFormatterShortStyle];
    
    //models
    wkService = [[WorkoutSessionService alloc]init];
    wkService.delegate = self;
    
    [wkService startWorkout];
}//eom

- (void)willActivate {
    // This method is called when watch view controller is about to be visible to user
    [super willActivate];
    
    
}//eom

- (void)didDeactivate {
    // This method is called when watch view controller is no longer visible
    [super didDeactivate];
}//eom


#pragma mark - Actions
- (IBAction)StartWorkout {
    [wkService startWorkout];
}//eo-a

- (IBAction)EndWorkout {
    [wkService endWorkout];
}//eo-a

-(void)updateUI:(BOOL)startedWorkout
{
//    if (startedWorkout) {
//        [startWkButton setHidden:true];
//        [endWkButton setHidden:false];
//    } else {
//        [startWkButton setHidden:false];
//        [endWkButton setHidden:true];
//    }
}//eom

#pragma mark - Service Delegates
-(void)WorkoutSessionServiceDidStartWorkoutAtDate:(NSDate *)date
{
    NSString * startedString = [dateF stringFromDate:date];
    startedString = [NSString stringWithFormat:@"Start: %@", startedString];
    
    //Update UI
    [wkStartDateLabel setText:startedString];
    [self updateUI:true];
    
    //TODO: notify phone
}//eom

-(void)WorkoutSessionServiceDidStopWorkoutAtDate:(NSDate *)date
{
    NSString * endedString = [dateF stringFromDate:date];
    endedString = [NSString stringWithFormat:@"End: %@", endedString];
    
    //update UI
    [wkEndDateLabel setText:endedString];
    [self updateUI:false];
    
    //TODO: notify phone
}//eom

-(void)WorkoutSessionServiceDidReceiveHeartrate:(double)hrValue
{
    NSString * hrStringValue = [NSString stringWithFormat:@"%.1f bpm", hrValue];
    [wkHrValueLabel setText:hrStringValue];
    
    //TODO: notify phone
}//eom

-(void)WorkoutSessionServiceErrorOccurred:(NSError *)error
{
    [wkStartDateLabel setText:[error localizedDescription]];
    [self updateUI:false];
}//eom

-(void)WorkoutSessionServiceAnotherSessionAlreadyStarted
{
    
    [wkStartDateLabel setText:@"Already"];
    [wkEndDateLabel setText:@"Started"];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [wkService startMonitoringHeartrate:[NSDate date]];
    });
    
    
    [startWkButton setBackgroundColor:[UIColor lightGrayColor]];
    
    //TODO: notify phone
}//eom

-(void)WorkoutSessionServiceConfigError:(NSError *)error
{
    [startWkButton setBackgroundColor:[UIColor darkGrayColor]];
}//eom


-(void)WorkoutSessionServiceHealthPermissionFailed:(NSError *)error
{
    [wkEndDateLabel setText:@"Failed Permission"];
    
    if (error != nil) {
        [wkEndDateLabel setText:[error localizedDescription]];
    }
    
    
    [startWkButton setBackgroundColor:[UIColor redColor]];
    
    //TODO: notify phone
}//eom

@end

File: watchOS2/tipCalculator/tipCalculator WatchKit Extension/InterfaceController.swift
//
//  InterfaceController.swift
//  tipCalculator WatchKit Extension
//
//  Created by Luis Castillo on 1/20/16.
//  Copyright © 2016 LC. All rights reserved.
//

import WatchKit
import Foundation


class InterfaceController: WKInterfaceController {
    
    @IBOutlet var inputAmountLabel: WKInterfaceLabel!
    
    @IBOutlet var decreaseAmountButton: WKInterfaceButton!
    
    @IBOutlet var increaseAmountButton: WKInterfaceButton!
    
    @IBOutlet var percent15Label: WKInterfaceLabel!
    @IBOutlet var percent18Label: WKInterfaceLabel!
    @IBOutlet var percent20Label: WKInterfaceLabel!
    
    var inputAmount:Double = 50.0
    
    //MARK: - View Loading
    override func awakeWithContext(context: AnyObject?) {
        super.awakeWithContext(context)
        
        // Configure interface objects here.
    }//eom
    
    override func willActivate()
    {
        // This method is called when watch view controller is about to be visible to user
        super.willActivate()
        
        self.updateLabelAmounts()
        
    }//eom
    
    override func didDeactivate()
    {
        // This method is called when watch view controller is no longer visible
        super.didDeactivate()
    }//eom
    
    //MARK: - Label Update
    func updateLabelAmounts()
    {
        let valueToDisplay = "$ \(inputAmount)"
        self.inputAmountLabel.setText(valueToDisplay)
        
        self.updatePercentValues()
    }//eom
    
    //MARK: - Calculations
    
    @IBAction func increaseAmount()
    {
        self.inputAmount += 2.5
        
        self.updateLabelAmounts()
    }//eo-a
    
    
    @IBAction func decreaseAmount()
    {
        inputAmount -= 2.5
        if inputAmount < 0
        {
            inputAmount = 0
        }
        
        self.updateLabelAmounts()
    }//eo-a
    
    
    func updatePercentValues()
    {
        var calcValue = 0.0
        var valueToDisplay = ""
        
        //15
        calcValue = inputAmount * 15 / 100
        valueToDisplay = "$ \(calcValue)"
        self.percent15Label.setText(valueToDisplay)
        
        //18
        calcValue = inputAmount * 18 / 100
        valueToDisplay = "$ \(calcValue)"
        self.percent18Label.setText(valueToDisplay)
        
        //20
        calcValue = inputAmount * 20 / 100
        valueToDisplay = "$ \(calcValue)"
        self.percent20Label.setText(valueToDisplay)
    }//eom

}
